servicetalk-client-api/src/main/java/io/servicetalk/client/api/ConnectionFactory.java

```diff
@@ -38,6 +42,32 @@
      * @param address to connect.
      * @param observer {@link TransportObserver} for the newly created connection.
      * @return {@link Single} that emits the created connection.
+     * @deprecated Use {@link #newConnection(Object, ContextMap, TransportObserver)}.
+     */
+    @Deprecated // FIXME: 0.43 - remove deprecated method
+    default Single<C> newConnection(ResolvedAddress address, @Nullable TransportObserver observer) {
+        throw new UnsupportedOperationException("ConnectionFactory#newConnection(ResolvedAddress, TransportObserver) " +
+                "is not supported by " + getClass());
+    }
+
+    /**
+     * Creates and asynchronously returns a connection.
+     *
+     * @param address to connect.
+     * @param context {@link ContextMap context} of the caller (e.g. request context) or {@code null} if no context
+     * provided. {@code null} context may also mean that a connection is created outside the normal request processing
+     * (e.g. health-checking).
+     * @param observer {@link TransportObserver} for the newly created connection or {@code null} if no observer
+     * provided.
+     * @return {@link Single} that emits the created connection.
      */
-    Single<C> newConnection(ResolvedAddress address, @Nullable TransportObserver observer);
+    default Single<C> newConnection(ResolvedAddress address, @Nullable ContextMap context,
+                                    @Nullable TransportObserver observer) { // FIXME: 0.43 - remove default impl
+        return Single.defer(() -> {
```

Comment:
What will the implementation be when Connection factory implementations override this method? Are they to implement the defer and the addition of the key to the context?

Reply:
No, the `defer` here is only necessary to "glue" the old overload with `DeprecatedToNewConnectionFactoryFilter`. If users implement this method, it's expected that they will invoke the same method in delegate chain, skipping the AsyncContext dance and directly propagating the `context` to the next filter.